File: java/11.ContainerWithMostWater.java
/*
Given n non-negative integers (a_1, a_2, ..., a_n)where each represents a point at coordinate (i, a_i).
‘ n ‘ vertical lines are drawn such that the two endpoints of line i is at (i, a_i) and (i, 0) . 
Find two lines, which together with x-axis forms a container, such that the container contains the most water.
*/
public class Solution {
    public int maxArea(int[] height) {
        int left = 0, right = height.length -1;
        int max = 0;
        while (left < right) {
            max = Math.max(max, Math.min(height[left], height[right]) * (right - left));
            if (height[left] < height[right]) {
                left ++;
            }
            else {
                right --;
            }
        }
        return max;
    }
}

File: py/11. Container With Most Water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        l, r, maxArea = 0, len(height) - 1, -1
        while l != r:
            maxArea = max(maxArea, (r - l) * min(height[l], height[r]))
            if height[l] < height[r]:
                l += 1
            else:
                r -= 1
        return maxArea
